src/raft/recv_install_snapshot.c

```diff
@@ -328,34 +279,482 @@ static const struct sm_conf leader_states[LS_NR] = {
+	(void)msg;
+	(void)state;
+	struct raft_message *reply;
+	reply = raft_malloc(sizeof(*reply));
+	assert(reply != NULL);
+	// TODO: construct proper message with the next range of checksums required.
+	reply->type = RAFT_IO_SIGNATURE;
+
+	return reply;
+}
+
+void process_snapshot_installation_cp_or_mv(struct snapshot_leader_state *state, const struct raft_message *msg) {
+	if (msg->type == RAFT_IO_INSTALL_SNAPSHOT_CP_RESULT) {
+		state->last_page_acked = msg->install_snapshot_cp_result.last_known_page_no;
+	} else if (msg->type == RAFT_IO_INSTALL_SNAPSHOT_MV_RESULT) {
+		state->last_page_acked = msg->install_snapshot_mv_result.last_known_page_no;
+	}
+}
+
+void send_mv_or_cp_cb(struct raft_io_send *req, int status) {
+	(void)status;
+
+	struct snapshot_leader_state *state = (struct snapshot_leader_state *)req->data;
+	sm_move(&state->sm, LS_SNAPSHOT_CHUNCK_SENT);
+	raft_free(req);
+}
+
+void async_send_mv_or_cp(struct snapshot_leader_state *state) {
+	(void)state;
+	struct raft_message *reply;
+	struct raft_io_send *reply_req;
+
+	reply = raft_malloc(sizeof(*reply));
+	assert(reply != NULL);
+	reply_req = raft_malloc(sizeof(*reply_req));
+	assert(reply_req != NULL);
+	reply_req->data = state;
+
+	// TODO: construct proper message.
+	reply->type = RAFT_IO_INSTALL_SNAPSHOT_CP;
+	state->io->async_send_message(reply_req, reply, send_mv_or_cp_cb);
+}
+
+void calculate_local_checksums(struct raft_io_async_work *req, int status) {
+	(void)status;
+	struct insert_checksum_data *data = (struct insert_checksum_data *)req->data;
+	struct snapshot_leader_state *state = data->state;
+
+	raft_free(req->data);
+	raft_free(req);
+	// TODO: Implement reading data.
+	sm_move(&state->sm, LS_SNAPSHOT_INSTALLATION_STARTED);
+	async_send_mv_or_cp(state);
+}
+
+void async_insert_checksums_calculate_local(struct snapshot_leader_state *state, const struct raft_message *msg) {
+	struct insert_checksum_data *data;
+	data = raft_malloc(sizeof(*data));
+	assert(data != NULL);
+	*data = (struct insert_checksum_data) {
+		.state = state,
+		.cs = msg->signature_result.cs,
+		.cs_nr = msg->signature_result.cs_nr,
+	};
+
+	struct raft_io_async_work *work;
+	work = raft_malloc(sizeof(*work));
+	assert(work != NULL);
+	*work = (struct raft_io_async_work) {
+		.work = insert_checksums,
+		.data = data,
+	};
+	state->io->async_work(work, calculate_local_checksums);
+}
+
+void leader_tick(struct sm *leader, const struct raft_message *msg)
+{
+	(void)leader_states;
+
+	assert(leader != NULL);
+	assert(msg != NULL);
+
+	struct snapshot_leader_state *state =
+		CONTAINER_OF(leader, struct snapshot_leader_state, sm);
+	struct snapshot_leader_io *io = state->io;
+	assert(state != NULL && io != NULL);
+
+	PRE(is_main_thread());
+	PRE(msg->server_id == state->follower_id);
+	// TODO: timeouts.
+	int leader_state = sm_state(leader);
+	if (leader_state == LS_FOLLOWER_ONLINE) {
```

Comment:
Suggestion: Case construction looks more idiomatic and natural here.

Reply:
I was using a switch statement and had some problems when omitting one `break;` and also when creating variables as `case` does not create a scope.